File: MessagePackReview/MessagePackReview.Tests/UnitTests.cs
using MessagePack;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace MessagePackReview.Tests
{
    [TestClass]
    public class UnitTests
    {
        [TestMethod]
        public void TestBackwardsCompatibility1()
        {
            // ARRANGE
            var person = new Person { Name = "John Smith", Address = "Missouri" };
            var bytes = LZ4MessagePackSerializer.Serialize(person);

            // ACT
            var personWithPhone = LZ4MessagePackSerializer.Deserialize<PersonWithPhone>(bytes);

            // ASSERT
            Assert.AreEqual(person.Name, personWithPhone.Name);
            Assert.AreEqual(person.Address, personWithPhone.Address);
            Assert.AreEqual(null, personWithPhone.Phone);
        }

        [TestMethod]
        public void TestBackwardsCompatibility2()
        {
            // ARRANGE
            var personWithPhone = new PersonWithPhone { Name = "John Smith", Address = "Missouri", Phone = "00723121211" };
            var bytes = LZ4MessagePackSerializer.Serialize(personWithPhone);

            // ACT
            var person = LZ4MessagePackSerializer.Deserialize<Person>(bytes);

            // ASSERT
            Assert.AreEqual(personWithPhone.Name, person.Name);
            Assert.AreEqual(personWithPhone.Address, person.Address);
        }
    }
}

File: MessagePackReview/MessagePack.FsharpReview.Tests/UnitTests.fs
namespace MessagePackFsharpReview.Tests

open Microsoft.VisualStudio.TestTools.UnitTesting
open MessagePack
open MessagePackFSharpReview.DomainTypes
open MessagePack.Resolvers
open MessagePack.FSharp

module UnitTests =

    [<TestClass>]
    type TestClass () =

        [<TestMethod>]
        member __.TestDUSerialization () =

            CompositeResolver.RegisterAndSetAsDefault(
                FSharpResolver.Instance,
                StandardResolver.Instance
            )

            let data = Foo 999
            let bin = MessagePackSerializer.Serialize(data)

            match MessagePackSerializer.Deserialize<UnionSample>(bin) with
            | Foo _ -> Assert.IsTrue(true);
            | Bar _ -> Assert.Fail();


